### PR TITLE
[📈⌨Analytics] Build Analytics - Week 1

### PR DESCRIPTION
**Pull request by qurben**
`build-analytics.Rmd` should probably be renamed to something sensible, but that requires renaming of other files.

https://chart-increasing-keyboard-analytics.github.io/software-analytics-book/build-analytics.html

### CODE REVIEW

File: build-analytics.Rmd [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+This is a US patent grant for a method of predicting software build errors. This patent is owned
+my Microsoft. Using logistic regression a prediction can be made on the probability of a build
+failing. Using this method build errors can be better anticipated, which decreases the time until
+the build works again.
+
+### @beller2017oops
+
+This paper explores data from Travis CI^[See https://travis-ci.org] on a large scale by analyzing
+2,640,825 build logs of Java and Ruby builds. It uses <span
+style="font-variant:small-caps">TravisTorrent</span> as a data source. It is found that the number
```
Comment (Inventitech) [inline build-analytics.Rmd]: Please use pure Rmd syntax.

File: build-analytics.Rmd [hunk 2]
```diff
@@ -0,0 +1,120 @@
+# Build analytics
+
+## Background
+
+When building a project from source code to executables everything should go smoothly. This is not
+always the case, a build can break for several reasons. This chapter will give an overview of
+research done on build scripts and continuous integration.
```
Comment (Inventitech) [inline build-analytics.Rmd]: Isn't "build scripts" unnecessarily narrow?

File: build-analytics.Rmd [hunk 3]
```diff
@@ -0,0 +1,120 @@
+# Build analytics
+
+## Background
+
+When building a project from source code to executables everything should go smoothly. This is not
```
Comment (Inventitech) [inline build-analytics.Rmd]: I like the idea of the introduction, but think this has to change a bit. "everything should go smoothly"? What's the connection between the sentences?

File: build-analytics.Rmd [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+always the case, a build can break for several reasons. This chapter will give an overview of
+research done on build scripts and continuous integration.
+
+<!--
+ - Define build
+ - Define context
+ -->
+
+## Research Questions
```
Comment (Inventitech) [inline build-analytics.Rmd]: Please cf. email to follow later today

File: build-analytics.Rmd [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+ * What are causes of a broken build?
+ * With which goals is continuous integration applied?
+ * What can be used to effectively fix a broken build?
+
+## Search Strategy
+
+Using the initial seed consisting of @bird2017predicting, @beller2017oops, @rausch2017empirical,
+@beller2017travistorrent, @pinto2018work, @zhao2017impact, @widder2018m and @hilton2016usage we
+used references and similar keywords to find new papers to analyze.
```
Comment (Inventitech) [inline build-analytics.Rmd]: "similar" is vague, unscientific and not replicable.

File: build-analytics.Rmd [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+## Study Selection
+
+Through this we found the following papers
```
Comment (Inventitech) [inline build-analytics.Rmd]: The papers will need to be analyzed per RQ. Maybe you still want to summarize all papers reviewed, then keeping a summary like this is nice to include!

File: build-analytics.Rmd (hunk 1 above)
Comment (qurben) [reply build-analytics.Rmd]: This is from the Rmd syntax book https://bookdown.org/yihui/bookdown/markdown-syntax.html Is there a better way to do smallcaps or should we not bother with smallcaps?

Comment (Inventitech) [reply build-analytics.Rmd]: Indeed, it even works in the PDF, see page 11 of https://bookdown.org/yihui/bookdown/bookdown.pdf.

File: build-analytics.Rmd (hunk 5 above)
Comment (demasterr) [reply build-analytics.Rmd]: It might be nice to give the keywords used per scientific research paper.

File: build-analytics.Rmd [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+## Summary of papers
+
+### @bird2017predicting
+
+This is a US patent grant for a method of predicting software build errors. This patent is owned
+my Microsoft. Using logistic regression a prediction can be made on the probability of a build
```
Comment (demasterr) [inline build-analytics.Rmd]: This patent is owned [b]y Microsoft

File: build-analytics.Rmd [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+languages. It notes that adoption of CI is often part of a reorganisation. It collected
+information on the differences before and after adoption of CI. There is also a survey amongst
+developers to learn about their experiences in adopting Travis CI.
+
+### @widder2018m
+
+This paper analyzes what factors have impact on abandonment of Travis. They find that increased
+build complexity reduces the chance of abandonment, but larger projects abandon at a higher rate
+and that a project's language has significant but varying effect. A surprising result is that
+metrics of configuration attempts and knowledge dispersion in the project don't affect the rate of
```
Comment (demasterr) [inline build-analytics.Rmd]: I think we should avoid using contractions. Please change `don't` to `do not`.

File: build-analytics.Rmd [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+abandonment.
+
+### @hilton2016usage
+
+This paper explores which CI system developers use, how developers use CI and why developers use
+CI. For this it analyzes data from Github, Travis CI and it conducts a developer survey. It finds
+that projects using CI release twice as oftern, accept pull requests faster and have developers
```
Comment (demasterr) [inline build-analytics.Rmd]: oftern --> often

File: build-analytics.Rmd [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+who are less worried about breaking the build.
+
+### @vassallo2017tale
+
+Discusses the difference in failures on continuous integration between open source software (OSS)
```
Comment (demasterr) [inline build-analytics.Rmd]: [This paper d]iscusses the difference (...)

File: build-analytics.Rmd [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+and industrial software projects. For this 349 Java OSS projects and 418 project from ING
+Nederland, a financial organisation.
```
Comment (demasterr) [inline build-analytics.Rmd]: I believe you should use `organization` here, as we use American English.

File: build-analytics.Rmd [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+Using the initial seed consisting of @bird2017predicting, @beller2017oops, @rausch2017empirical,
+@beller2017travistorrent, @pinto2018work, @zhao2017impact, @widder2018m and @hilton2016usage we
+used references and similar keywords to find new papers to analyze.
+
+## Study Selection
+
+Through this we found the following papers
+
+## Summary of papers
```
Comment (chadepl) [inline build-analytics.Rmd]: Summaries are useful to get an idea of the paper, but I think they should be more consistent in the sense that they should be similarly structured. Verify that the summaries include the papers' research questions and their corresponding answers.

File: 02-build-analytics.Rmd [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+investigated which tools are being used, what types of issues make the build fail or raise
+warnings and how is responded to broken builds. 
+
+## What is the current state of practice in the field of build analytics?
+
+In this section, I will examine scientific papers to analyse the current trend of build analytics in the software development industry. 
+
+### @fowler2006continuous
+
+In this paper, Martin talks about the current state of the software industry in terms of Continuous Integration (CI) and comments on the practises required to implement CI effectively. He talks about his experience working for a large English electronics company where the development of a project took two years and the integration process took several months. Integration is a long and unpredictable process. Martin suggested this approach and that the two most common reactions he got were: “it can’t work (here)” or “doing it won’t make much difference”.  He expresses that most engineers don’t know how simple the process can be of setting the CI framework up. In this way, we get a glimpse into the practises popular within the industry regarding build analytics.
```
Comment (qurben) [inline 02-build-analytics.Rmd]: @GautamMiglani Please wrap your lines.

File: book.bib [hunk 14]
```diff
@@ -114,4 +114,22 @@ @inproceedings{zampetti2017open
...
+@article{fowler2006continuous,
+  title={Continuous integration},
+  author={Fowler, Martin and Foemmel, Matthew},
+  journal={Thought-Works) http://www. thoughtworks. com/Continuous Integration. pdf},
+  volume={122},
+  pages={14},
+  year={2006}
+}
+
+@inproceedings{stolberg2009enabling,
```
Comment (qurben) [inline book.bib]: Please document how you found these papers.

File: build-analytics.Rmd (hunk 2 above)
Comment (qurben) [reply build-analytics.Rmd]: Fixed in #12

File: build-analytics.Rmd (hunk 4 above)
Comment (qurben) [reply build-analytics.Rmd]: Fixed in #12

**Comment by Inventitech** [comment]
Happy to merge once the file is renamed to `03-build-analytics.Rmd` (and rebased).

File: 03-build-analytics.Rmd [hunk 15]
```diff
@@ -1,44 +1,152 @@
...
-    - Tools and companies creating / employing them
-    - Case studies and their findings
+<!-- Subject to change -->
 
-* **RQ3** Open challenges and future research required
+ * What are causes of a broken build?
+ * With which goals is continuous integration applied?
+ * What can be used to effectively fix a broken build?
 
-## Research protocol
```
Comment (Inventitech) [inline 03-build-analytics.Rmd]: What about sticking to the general term here instead of your own? :-)

**Comment by Inventitech** [comment]
1 comment, plus 1 conflict

**Comment by qurben** [comment]
@Inventitech resolved conflicts and comment. The latter part is being build to the proper chapter structure in #12.

**Pull request merged**